### PR TITLE
Extract in/out HID report sizes from descriptors

### PR DESCRIPTION
**Pull request by FabianHenneke**
The U2F/CTAP specifications for the HID transport allow arbitrary
sizes for the HID input/output reports, not just the standard 64 bytes.
This commit adds the general logic needed to deal with varying report
sizes and implements the relevant descriptor parsing for Linux.

### CODE REVIEW

File: src/consts.rs [hunk 1]
```diff
@@ -5,17 +5,17 @@
 // Allow dead code in this module, since it's all packet consts anyways.
 #![allow(dead_code)]
 
-pub const HID_RPT_SIZE: usize = 64;
```
Comment (jcjones) [inline src/consts.rs:8]: This is used by all the other platforms as well, so we'll have to find/replace across the repo.

**Comment by FabianHenneke** [comment]
> Thank you for this, @FabianHenneke! Can you give me an example of a device that provides variable size HID reports?

Dynamic or at least non-standard HID report sizes are used by authenticators that implement the HID protocol via Bluetooth. I am the developer of such an authenticator, [WearAuthn](https://github.com/FabianHenneke/WearAuthn], and I know that @rstoica is working on such an authenticator due to #32. While Bluetooth as a HID transport is not part of the FIDO specs, it is supported by the Windows WebAuthn API and by Chrome on Linux, macOS and Chrome OS. It would be great if Firefox could also become compatible on non-Windows platforms.

I will take another look at the PR and take care of the HID_RPT_SIZE issue.

**Comment by jcjones** [comment]
> > Thank you for this, @FabianHenneke! Can you give me an example of a device that provides variable size HID reports?
> 
> Dynamic or at least non-standard HID report sizes are used by authenticators that implement the HID protocol via Bluetooth. I am the developer of such an authenticator, [WearAuthn](https://github.com/FabianHenneke/WearAuthn), and I know that @rstoica is working on such an authenticator due to #32. While Bluetooth as a HID transport is not part of the FIDO specs, it is supported by the Windows WebAuthn API and by Chrome on Linux, macOS and Chrome OS. It would be great if Firefox could also become compatible on non-Windows platforms.

That seems very doable. I just want to be able to test if I can. I'll see about getting something running WearOS if I can.

> I will take another look at the PR and take care of the HID_RPT_SIZE issue.

Thanks. I put a skeleton fix in 5d284842c9e58aa3e578bc2a1716333d70248593 - I spent a short while this morning trying to figure out how to get the same data out of iokit but it needs more research than I could pull off at the time.

**Comment by rstoica** [comment]
> > > Thank you for this, @FabianHenneke! Can you give me an example of a device that provides variable size HID reports?
> > 
> > 
> > Dynamic or at least non-standard HID report sizes are used by authenticators that implement the HID protocol via Bluetooth. I am the developer of such an authenticator, [WearAuthn](https://github.com/FabianHenneke/WearAuthn), and I know that @rstoica is working on such an authenticator due to #32. While Bluetooth as a HID transport is not part of the FIDO specs, it is supported by the Windows WebAuthn API and by Chrome on Linux, macOS and Chrome OS. It would be great if Firefox could also become compatible on non-Windows platforms.
> 
> That seems very doable. I just want to be able to test if I can. I'll see about getting something running WearOS if I can.
> 
If Android OS would be easier to get a hang on for the necessary testing you are welcome to try [WioKey](https://www.wiokey.de/en/#download) (Early Access) as well. Source code is not yet out, but it uses the same underlying Bluetooth HID within the Android Bluetooth stack.
> > I will take another look at the PR and take care of the HID_RPT_SIZE issue.
> 
> Thanks. I put a skeleton fix in [5d28484](https://github.com/mozilla/authenticator-rs/commit/5d284842c9e58aa3e578bc2a1716333d70248593) - I spent a short while this morning trying to figure out how to get the same data out of iokit but it needs more research than I could pull off at the time.

**Comment by jcjones** [comment]
> If Android OS would be easier to get a hang on for the necessary testing you are welcome to try [WioKey](https://www.wiokey.de/en/#download) (Early Access) as well. Source code is not yet out, but it uses the same underlying Bluetooth HID within the Android Bluetooth stack.

Thanks - a good idea. My general Android test device (a Samsung S7 edge) is apparently incompatible, but I'll give it a try on my other devices tomorrow.

I don't have to have running code to finish this review and accept. Would just be nice to be able to work it out for MacOS and Windows, too! I should have this patch into the next release of `authenticator-rs`, _probably_ in Firefox 80.

**Comment by rstoica** [comment]
> Thanks - a good idea. My general Android test device (a Samsung S7 edge) is apparently incompatible, but I'll give it a try on my other devices tomorrow.

Yes, unfortunately not all of the devices have the Bluetooth HID profile enabled, but most of the Android 9+ should support it. The other Bluetooth LE GATT solution is currently blacklisted by Google.

> I don't have to have running code to finish this review and accept. Would just be nice to be able to work it out for MacOS and Windows, too! I should have this patch into the next release of `authenticator-rs`, _probably_ in Firefox 80.

Current code version (without this PR) works out of the box on Windows machines as it goes through Windows Hello and the drivers pick up the Bluetooth automatically as an input device anyway. I don't have now a MacOS to check particularly the Mozilla browser, but I can do it maybe in the next days. I know though that it works on Mac with other browser clients as another colleague tested it already, but don't recall exactly the state for Firefox and don't want to pass misinformation. I will update here this comment once I find out.

**Comment by FabianHenneke** [comment]
> Yes, unfortunately not all of the devices have the Bluetooth HID profile enabled, but most of the Android 9+ should support it. The other Bluetooth LE GATT solution is currently blacklisted by Google.

It also doesn't work with OnePlus and (most) Nokia phones. More recent Samsung phones should work, but Pixels are ideal.

> Current code version (without this PR) works out of the box on Windows machines as it goes through Windows Hello and the drivers pick up the Bluetooth automatically as an input device anyway. I don't have now a MacOS to check particularly the Mozilla browser, but I can do it maybe in the next days. I know though that it works on Mac with other browser clients as another colleague tested it already, but don't recall exactly the state for Firefox and don't want to pass misinformation. I will update here this comment once I find out.

I have added support for dynamic HID sizes on macOS to libfido2 in https://github.com/Yubico/libfido2/pull/180 (and Windows in https://github.com/Yubico/libfido2/pull/182, but as @rstoica said, Windows Hello makes this rather uninteresting). Maybe that code can be helpful, as I may not have time to work on it myself.

**Comment by jcjones** [comment]
As an update, I should be able to take this for the next Firefox cycle, starting end-of-month.

**Pull request merged**